### DISCUSSION TITLE
Pthreads no longer allow user to have custom instantiateWasm function

### DISCUSSION THREAD

**Discussion by shinji-TH** [emscripten-core/emscripten#22905] [Q&A] [answered]
Hi @sbc100 , 
I'm upgrading my app from EMSDK 3.1.41 and noticed that in #21701  , pthreads now have their own `instantiateWasm` functions. 
https://github.com/emscripten-core/emscripten/blob/e6ed9d3fc2821a72cc8ded4a02cac72b9f306f88/src/runtime_pthread.js#L68

My app uses our own `instantiateWasm` to perform initialization logic for the main threads and worker threads. 
Hence wishes to check if this change is as expected and if there are any recommendations to have worker threads to initialize any custom code before calling `var instance = new WebAssembly.Instance(module, getWasmImports());`

**Comment by sbc100** [comment]
Interesting.   Even before #21701 we still used the `instantiateWasm` method in workers to load the wasm file: 
 https://github.com/emscripten-core/emscripten/blob/9f19ec9eef86bd6301ce02fde3aa4c3b5a213cc6/src/worker.js#L97-L114

I wonder if the different is that this used to come before the preJS files and now is comes after?  Or some kind ordering thing like that?  How are you injecting your `instantiateWasm`?

**Reply by sbc100** [reply]
It seems like we were already depending on the `instantiateWasm` from `worker.js` being used since it was setting up the `wasmPromiseResolve` promise.   Was you custom `instantiateWasm` also doing this?

**Reply by shinji-TH** [reply]
Yup, the instantiateWasm function was there previously. 
However when we call `loadWasmModuleToWorker,` we overwrite the function by providing the worker with the Module of the main script.
https://github.com/emscripten-core/emscripten/blob/9f19ec9eef86bd6301ce02fde3aa4c3b5a213cc6/src/library_pthread.js#L381
Hence, during initialization, the worker threads run with the user's version of `instantiateWasm` instead of the worker.js copy.
In the new change, since there was no need for worker.js, `instantiateWasm` was loaded with the threads version as it is packed in as the last in the javascript.

**Reply by sbc100** [reply, marked as answer]
I just landed a change the re-introduces `instantiateWasm` in pthread workers: #22912

However, even after this change I fear that #21701 may have change pthread creations in other ways that break your use case since `urlOrBlob` is not longer passed to new workers.   `mainScriptUrlOrBlob` can still be used to tell emscripten how to find main script, but this is then used directly to create new workers rather then passed as an argument to the separate `worker.js` file (which no longer exists).

Perhaps you can say more about what you are trying to do in your `instantiateWasm` withing your pthread workers?  Pthread workers already have access to the wasm module directly so they should know how to instantiate it.

**Reply by shinji-TH** [reply]
We are using `instantiateWasm` for a few things
- Initialization of WebGPU
- Creation of offscreen canvas. 
- initialization of service worker channel 

Thanks for helping to land the changes!

**Reply by sbc100** [reply]
I see, but none of those actually effect instantiation of the module on each thread do they?   Are thos per-thread startup tasks that you want to perform on each worker as it starts up?

**Reply by shinji-TH** [reply]
Yes, those are per-thread startup tasks due to our legacy code, where we render within the thread worker, as it is faster than proxy calls in our previous profiling.

**Reply by sbc100** [reply]
And then how does your version of `instantiateWasm` actually perform the task of instantiation?  Where does it find the wasm module?

Is the work you need to do, per-pthread or per-worker?   I'm hoping we can give you some other hooks since I don't think `instantiateWasm` sounds like the right place to do that work you are talking about here.

**Reply by shinji-TH** [reply]
Yes, having new hooks would be nice. 
What is the difference between per-threads and per-worker?

Currently, the work is done via per-worker.
Our C++ code calls WebGPU functions within the thread function, which requires the workers to have access to WebGPU devices and offscreen canvas

For each worker:
- Under  the Initialization of WebGPU
    we call `navigator.gpu.requestAdapter()` to request for the device and assign the device under `Module["preinitializedWebGPUDevice"]` which allows `_emscripten_webgpu_get_device()`
- Under the creation of offscreen canvas 
  we modify`Module['mainScriptUrlOrBlob']` with search params by providing the canvas width and height to the worker via `urlOrBlob`.
  With the width and height, we create an offscreen canvas via 
  
>        Module.canvas = new OffscreenCanvas(cWidth, Height);
>        Module.stagingCanvas = new OffscreenCanvas(cWidth, cHeight);
>        Module.stagingContext = Module.stagingCanvas.getContext("2d", { willReadFrequently: true });
>        Module.requestedRenderMode = renderMode;
>        specialHTMLTargets["!canvas"] = Module.canvas;

which WebGPU functions render onto. Hence without `urlOrBlob ` we are unable to pass values over to the worker.
Is there any recommendation to resolve this issue?

**Reply by sbc100** [reply]
The difference between per-worker and per-pthread is that workers can be re-used for multiple pthreads.  Emscripten caches the workers it creates in order to speed up the creation of new threads.   It sounds like the work you want to do is basically per-worker, but would also be fine if it was done per-pthread.

In terms of a solution could you do in your C/C++ as part of the thread entry point?   You can use `emscripten_get_canvas_size` on worker threads to get the canvas size.

e.g.

```
EM_JS(void, createOffscreenCanvas, (int canvasWidth, int canvasHeight), {
  .. 
});

int thread_main() {
   int width;
   int height;
   int isFullscreen;
   emscripten_get_canvas_size(&width, &height, &isFullscreen);
   createOffscreenCanvas(width, height);
   ... 
}
```

I think you could do the same for `requestAdapter` although you would have to deal with the fact that its async.

**Reply by sbc100** [reply]
Actually I think you can use use `Module['preInit']`.  That just confirmed that that callback will run on each thread.

If you need to do async work in that function that should occur before the thread starts you can use `addRunDependency` and `removeRunDependency` when you are done.

**Reply by shinji-TH** [reply]
Thanks for the recommendation!! 
Initializing from C/C++ from thread entry point might not be feasible on my end, as we have multiple developers who use standard threads when required. There is no centralized thread entry point without any huge code changes. 
`Module['preInit']` seems more feasible to implement on each worker. 'Module['mainScriptUrlOrBlob ']` still seems to be the only way to pass values/flags from Main worker to the remaining thread workers.

**Reply by sbc100** [reply]
Yes, unfortunately `mainScriptUrlOrBlob` is not longer sent to worker since the works is started already running that script so it doesn't need that information.  

So for `requestAdapter` is think `preInit` is a good option (since its async and you can use `addRunDependency` to delay thread startup until its done).

For `new OffscreenCanvas` I think you need to hook into thread startup somehow and then call `emscripten_get_canvas_size` to get the canvas size.  I wonder if you could use some kind of static TLS constructor.   Something like this mabye:

```
struct CanvasCreator {
   CanvasCreator() {
      EM_JS({
          if (Module.stagingCanvas) return; // already run on this worker 
          ...
      });
   }
};
thread_local CanvasCreator c;
```

**Reply by sbc100** [reply]
Actually I think the static constructor approach wont work since I think the object are created lazily not when a thread starts.

**Reply by shinji-TH** [reply]
`mainScriptUrlOrBlob` still seems to work for me when I modify the field in the root of the script (--pre-js). The main thread would load the script and change the field before initializing workers for pthreads. Is that not intended? 
https://github.com/emscripten-core/emscripten/blob/015b4004fe49a1b9fc6bf070ffcb93d1cfd93d44/src/library_pthread.js#L439

**Reply by sbc100** [reply]
yes `mainScriptUrlOrBlob` is used to create the workers, but it is no longer passed to the worker via the `urlOrBlob` property.   If that is enough for your use case that is great!

**Comment by sbc100** [comment]
One more clarifying question:  What are you trying to do in the pthreads' instantiateWasm is that is different from simply instantiating the recieved module?   Are you using the module that is received via postMessage (previously referred to as `Module['wasmModule']`) or are you retrieving it some other way?
